{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c941ac39-c7df-4fc8-95a3-0d68d6b86b71",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a3409888-5741-49af-9fd2-25a83833aeb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function to allow users to choose how many samples(AD and AR) they need to simulate.\n",
    "def pairs_variants(n, m):\n",
    "    \"\"\"\n",
    "    Simulate causative variants of paired samples for AD (Autosomal Dominant) and AR (Autosomal Recessive) inheritance.\n",
    "\n",
    "    Args:\n",
    "        n (int): A non-negative, even integer specifying the number of Autosomal Dominant (AD) samples to simulate.\n",
    "        m (int): A non-negative, even integer specifying the number of Autosomal Recessive (AR) samples to simulate.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves two text files:\n",
    "            1. `pairs_variants.txt`: Contains the simulated causative variants for each sample.\n",
    "            2. `pairs_chrom_pos.txt`: Contains the chromosome and position information for the causative variants.\n",
    "\n",
    "    \"\"\"\n",
    "    # Read the file into a DataFrame, assuming it is tab-separated\n",
    "    df = pd.read_csv('variants_pairs_vcf.csv')\n",
    "    \n",
    "    # Ensure n and m are within the specified limits\n",
    "    n = min(n, 1184)\n",
    "    m = min(m, 1410)\n",
    "    \n",
    "    # Extract first n rows and last m rows\n",
    "    rows_extracted = pd.concat([df.head(n), df.tail(m)])\n",
    "    \n",
    "    # Extract first 5+n columns and last m columns\n",
    "    columns_extracted = pd.concat([df.iloc[:, :5+n], df.iloc[:, -m:]], axis=1)\n",
    "    \n",
    "    # Combine extracted rows and columns\n",
    "    extracted_df = rows_extracted[columns_extracted.columns]\n",
    "    \n",
    "    # Save the result to a txt file\n",
    "    extracted_df.to_csv('pairs_variants.txt', sep='\\t', index=False, header=False)\n",
    "    \n",
    "    # Extract the first two columns (CHROM and POS)\n",
    "    chrom_pos_columns = extracted_df[['CHROM', 'POS']]\n",
    "\n",
    "    # Save to a text file\n",
    "    chrom_pos_columns.to_csv(\"pairs_chrom_pos.txt\", index=False, header=False, sep='\\t')\n",
    "    \n",
    "    return extracted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "46561cf5-756a-449e-b7fd-13c09f149faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function to allow users to get the causative genes for the simulated patients.\n",
    "def pairs_genes(n, m): \n",
    "    \"\"\"\n",
    "    Simulate causative genes of paired samples for AD (Autosomal Dominant) and AR (Autosomal Recessive) inheritance.\n",
    "\n",
    "    Args:\n",
    "        n (int): A non-negative, even integer specifying the number of Autosomal Dominant (AD) samples to simulate.\n",
    "        m (int): A non-negative, even integer specifying the number of Autosomal Recessive (AR) samples to simulate.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves a csv file:\n",
    "        `pairs_genes.csv`: Contains the simulated causative genes for each sample.\n",
    "        \n",
    "    \"\"\"    \n",
    "    # Read the file into a DataFrame, assuming it is tab-separated\n",
    "    df = pd.read_csv('pairs_variants_gene_orpha_inher.csv')\n",
    "    \n",
    "    # Ensure n and m are within the specified limits\n",
    "    n = min(n, 1184)\n",
    "    m = min(m, 1410)\n",
    "    \n",
    "    # Extract first n rows and last m rows\n",
    "    rows_extracted = pd.concat([df.head(n), df.tail(m)])\n",
    "    \n",
    "    # Extract first 5+n columns and last m columns\n",
    "    columns_extracted = pd.concat([df.iloc[:, :5+n], df.iloc[:, -m:]], axis=1)\n",
    "    \n",
    "    # Combine extracted rows and columns\n",
    "    extracted_df = rows_extracted[columns_extracted.columns]\n",
    "    \n",
    "    # Save the result to a txt file\n",
    "    extracted_df.to_csv('pairs_genes.csv', index=False)\n",
    "    \n",
    "    return extracted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2a1f5522-3dfb-4cd8-b719-8d71c12e702a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function allowing users to get the phenotypes for the simulated patients.\n",
    "def pairs_phenotypes(n, m):\n",
    "    \n",
    "    \"\"\"\n",
    "    Simulate phenotypes of paired samples for AD (Autosomal Dominant) and AR (Autosomal Recessive) inheritance.\n",
    "\n",
    "    Args:\n",
    "        n (int): A non-negative, even integer specifying the number of Autosomal Dominant (AD) samples to simulate.\n",
    "        m (int): A non-negative, even integer specifying the number of Autosomal Recessive (AR) samples to simulate.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves a csv file:\n",
    "        `pairs_phenotypes.csv`: Contains the simulated phenotypes for each sample.\n",
    "        \n",
    "    \"\"\"    \n",
    "    \n",
    "    # Read the file into a DataFrame, assuming it is tab-separated\n",
    "    df = pd.read_csv('pairs_orphacode_hposet.csv')\n",
    "    \n",
    "    # Ensure n and m are within the specified limits\n",
    "    n = min(n, 1184)\n",
    "    m = min(m, 1410)\n",
    "    \n",
    "    # Extract first n rows and last m rows\n",
    "    rows_extracted = pd.concat([df.head(n), df.tail(m)])\n",
    "    \n",
    "    # Save the result to a txt file\n",
    "    rows_extracted.to_csv('pairs_phenotypes.csv')\n",
    "    \n",
    "    print(rows_extracted.head())\n",
    "    print(\"File saved as 'pairs_phenotypes.csv'\")\n",
    "    \n",
    "    return rows_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c5244cb4-a5f4-4413-8b91-c00e15024391",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Function to randomly select HPO terms to simulate different scenarios for phenotype simulation\n",
    "def pairs_hpo(i, n, m):\n",
    "\n",
    "    \"\"\"\n",
    "    Simulate user-defined scenarios by selecting a specified number of HPO terms related to rare diseases \n",
    "    for phenotype simulation.\n",
    "\n",
    "    Args:\n",
    "        i (int): A non-negative integer specifying the number of rare disease-related HPO terms to select, \n",
    "        simulating scenarios where 'i' rare disease-related HPO terms are randomly chosen for phenotype simulation.  \n",
    "        \n",
    "        n (int): A non-negative integer specifying the number of rare disease-related HPO terms to select, \n",
    "        simulating scenarios where 'n' rare disease-related HPO terms are randomly chosen. Additionally, \n",
    "        'm' HPO terms are randomly selected from the entire set of HPO terms to introduce noise for phenotype simulation.\n",
    "        \n",
    "        m (int): A non-negative integer specifying the number of rare disease-unrelated HPO terms to select, \n",
    "        simulating scenarios where 'n' rare disease-related HPO terms are randomly chosen. Additionally, \n",
    "        'm' HPO terms are randomly selected from the entire set of HPO terms to introduce noise for phenotype simulation.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves a csv file:\n",
    "        `pairs_hpo.csv`: Contains the simulated HPO terms user selected for each sample.\n",
    "        \n",
    "    \"\"\"    \n",
    "        \n",
    "    # read the orphanet code and hpoid of pairs information\n",
    "    df_pairs_hposet = pd.read_csv('pairs_phenotypes.csv')\n",
    "    \n",
    "    def random_hpo(hpo_str, i):\n",
    "        # Split the HPO terms by \"; \" and randomly select n terms\n",
    "        hpo_terms = hpo_str.split(\"; \")\n",
    "        selected_terms = random.sample(hpo_terms, min(i, len(hpo_terms)))  # limit to available terms\n",
    "        return \"; \".join(selected_terms)\n",
    "    \n",
    "    # Function to select random HPO terms from a string\n",
    "    def random_hpo_noise(hpo_str, num_terms):\n",
    "        # Split the HPO terms by \"; \" and randomly select the specified number of terms\n",
    "        hpo_terms = hpo_str.split(\"; \")\n",
    "        selected_terms = random.sample(hpo_terms, min(num_terms, len(hpo_terms)))  # limit to available terms\n",
    "        return selected_terms\n",
    "   \n",
    "    # read all the hpo term\n",
    "    df_all_hpo = pd.read_csv('hpo_id_all.csv')\n",
    "    \n",
    "    # Convert df_all_hpo 'hpo_id' column to a list of all HPO terms\n",
    "    all_hpo_terms = df_all_hpo['hpo_id'].tolist()\n",
    "    \n",
    "    # Apply the function to create the HPO_NOISE column in df_pairs_hposet\n",
    "    def combine_hpo_noise(row):\n",
    "        # Select n terms from the HPO_ALL column in df_pairs_hposet\n",
    "        selected_hpo_all = random_hpo_noise(row['HPO_ALL'], n)\n",
    "        # Select m terms from the df_all_hpo HPO list\n",
    "        selected_hpo_noise = random.sample(all_hpo_terms, min(m, len(all_hpo_terms)))\n",
    "        # Combine and return as a single string with \"; \" delimiter\n",
    "        return \"; \".join(selected_hpo_all + selected_hpo_noise)\n",
    "\n",
    "    # Apply the function to each row in df_pairs_hposet\n",
    "    df_pairs_hposet[f'HPO_{i}'] = df_pairs_hposet['HPO_ALL'].apply(lambda x: random_hpo(x, i))    \n",
    "\n",
    "    # Add the new column HPO_NOISE to df_pairs_hposet\n",
    "    df_pairs_hposet['HPO_NOISE'] = df_pairs_hposet.apply(combine_hpo_noise, axis=1)\n",
    "    \n",
    "    # Save the result to a csv file\n",
    "    df_pairs_hposet.to_csv('pairs_hpo.csv')\n",
    "\n",
    "    print(df_pairs_hposet.head())\n",
    "    print(\"File saved as 'pairs_hpo.csv'\")\n",
    "    \n",
    "    return df_pairs_hposet"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
