{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c941ac39-c7df-4fc8-95a3-0d68d6b86b71",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a3409888-5741-49af-9fd2-25a83833aeb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function to allow users to choose how many samples(AD and AR) they need to simulate.\n",
    "def pathway_variants(n, m):\n",
    "    \"\"\"\n",
    "    Simulate causative variants of samples for AD (Autosomal Dominant) and AR (Autosomal Recessive) inheritance.\n",
    "\n",
    "    Args:\n",
    "        n (int): A non-negative integer specifying the number of Autosomal Dominant (AD) samples to simulate.\n",
    "        m (int): A non-negative integer specifying the number of Autosomal Recessive (AR) samples to simulate.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves two text files:\n",
    "            1. `pathway_variants.txt`: Contains the simulated causative variants for each sample.\n",
    "            2. `pathway_chrom_pos.txt`: Contains the chromosome and position information for the causative variants.\n",
    "\n",
    "    \"\"\"\n",
    "        \n",
    "    # Read the file into a DataFrame, assuming it is tab-separated\n",
    "    df = pd.read_csv('variants_pathway_vcf.csv')\n",
    "    \n",
    "    # Ensure n and m are within the specified limits\n",
    "    n = min(n, 1026)\n",
    "    m = min(m, 1182)\n",
    "    \n",
    "    # Extract first n rows and last m rows\n",
    "    rows_extracted = pd.concat([df.head(n), df.tail(m)])\n",
    "    \n",
    "    # Extract first 5+n columns and last m columns\n",
    "    columns_extracted = pd.concat([df.iloc[:, :9+n], df.iloc[:, -m:]], axis=1)\n",
    "    \n",
    "    # Combine extracted rows and columns\n",
    "    extracted_df = rows_extracted[columns_extracted.columns]\n",
    "    \n",
    "    # Save the result to a txt file\n",
    "    extracted_df.to_csv('pathway_variants.txt', sep='\\t', index=False, header=False)\n",
    "    \n",
    "    # Extract the first two columns (CHROM and POS)\n",
    "    chrom_pos_columns = extracted_df[['CHROM', 'POS']]\n",
    "\n",
    "    # Save to a text file\n",
    "    chrom_pos_columns.to_csv(\"pathway_chrom_pos.txt\", index=False, header=False, sep='\\t')\n",
    "    \n",
    "    return extracted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "46561cf5-756a-449e-b7fd-13c09f149faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function to allow users to get the causative genes for the simulated patients.\n",
    "def pathway_genes(n, m): \n",
    "    \"\"\"\n",
    "    Simulate causative genes of samples for AD (Autosomal Dominant) and AR (Autosomal Recessive) inheritance.\n",
    "\n",
    "    Args:\n",
    "        n (int): A non-negative, even integer specifying the number of Autosomal Dominant (AD) samples to simulate.\n",
    "        m (int): A non-negative, even integer specifying the number of Autosomal Recessive (AR) samples to simulate.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves a csv file:\n",
    "        `pathway_genes.csv`: Contains the simulated causative genes for each sample.\n",
    "        \n",
    "    \"\"\"    \n",
    "    # Read the file into a DataFrame, assuming it is tab-separated\n",
    "    df = pd.read_csv('pathway_variants_gene_orpha_inher.csv')\n",
    "    \n",
    "    # Ensure n and m are within the specified limits\n",
    "    n = min(n, 1026)\n",
    "    m = min(m, 1182)\n",
    "     \n",
    "    # Extract first n rows and last m rows\n",
    "    rows_extracted = pd.concat([df.head(n), df.tail(m)])\n",
    "    \n",
    "    # Extract first 5+n columns and last m columns\n",
    "    columns_extracted = pd.concat([df.iloc[:, :5+n], df.iloc[:, -m:]], axis=1)\n",
    "    \n",
    "    # Combine extracted rows and columns\n",
    "    extracted_df = rows_extracted[columns_extracted.columns]\n",
    "    \n",
    "    # Save the result to a txt file\n",
    "    extracted_df.to_csv('pathway_genes.csv', index=False)\n",
    "    \n",
    "    return extracted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3c2e1722-2be3-4d87-9e51-ff27c2c4d098",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>ORPHACODE</th>\n",
       "      <th>INHERITANCE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>1806104</td>\n",
       "      <td>16330</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>FGFR3</td>\n",
       "      <td>15</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>1805658</td>\n",
       "      <td>2664079</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>FGFR3</td>\n",
       "      <td>15</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>66727442</td>\n",
       "      <td>13350</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>15</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>138391446</td>\n",
       "      <td>1344704</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ATP6V0A4</td>\n",
       "      <td>18</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>71185243</td>\n",
       "      <td>12228</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ATP6V1B1</td>\n",
       "      <td>18</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2203</th>\n",
       "      <td>17</td>\n",
       "      <td>8076766</td>\n",
       "      <td>929264</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>TMEM107</td>\n",
       "      <td>542310</td>\n",
       "      <td>AR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2204</th>\n",
       "      <td>4</td>\n",
       "      <td>108852800</td>\n",
       "      <td>989059</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>CYP2U1</td>\n",
       "      <td>542310</td>\n",
       "      <td>AR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2205</th>\n",
       "      <td>1</td>\n",
       "      <td>150990382</td>\n",
       "      <td>932732</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PRUNE1</td>\n",
       "      <td>544469</td>\n",
       "      <td>AR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2206</th>\n",
       "      <td>1</td>\n",
       "      <td>150990970</td>\n",
       "      <td>427231</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>PRUNE1</td>\n",
       "      <td>544469</td>\n",
       "      <td>AR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2207</th>\n",
       "      <td>6</td>\n",
       "      <td>132129383</td>\n",
       "      <td>1699972</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENPP1</td>\n",
       "      <td>544469</td>\n",
       "      <td>AR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2208 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS       ID REF ALT  GENEINFO  ORPHACODE INHERITANCE\n",
       "0         4    1806104    16330   G   T     FGFR3         15          AD\n",
       "1         4    1805658  2664079   C   G     FGFR3         15          AD\n",
       "2        15   66727442    13350   T   C    MAP2K1         15          AD\n",
       "3         7  138391446  1344704   T   C  ATP6V0A4         18          AD\n",
       "4         2   71185243    12228   T   C  ATP6V1B1         18          AD\n",
       "...     ...        ...      ...  ..  ..       ...        ...         ...\n",
       "2203     17    8076766   929264   G   C   TMEM107     542310          AR\n",
       "2204      4  108852800   989059   A   C    CYP2U1     542310          AR\n",
       "2205      1  150990382   932732   T   C    PRUNE1     544469          AR\n",
       "2206      1  150990970   427231   C   A    PRUNE1     544469          AR\n",
       "2207      6  132129383  1699972   A   T     ENPP1     544469          AR\n",
       "\n",
       "[2208 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('pathway_variants_gene_orpha_inher.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d20ef8a9-fa80-4da9-bc86-9f82bc958e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pathway_genes = pathway_genes(30, 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2a1f5522-3dfb-4cd8-b719-8d71c12e702a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function allowing users to get the phenotypes for the simulated patients.\n",
    "def pathway_phenotypes(n, m):\n",
    "    \n",
    "    \"\"\"\n",
    "    Simulate phenotypes of samples for AD (Autosomal Dominant) and AR (Autosomal Recessive) inheritance.\n",
    "\n",
    "    Args:\n",
    "        n (int): A non-negative, even integer specifying the number of Autosomal Dominant (AD) samples to simulate.\n",
    "        m (int): A non-negative, even integer specifying the number of Autosomal Recessive (AR) samples to simulate.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves a csv file:\n",
    "        `pathway_phenotypes.csv`: Contains the simulated phenotypes for each sample.\n",
    "        \n",
    "    \"\"\"    \n",
    "    \n",
    "    # Read the file into a DataFrame, assuming it is tab-separated\n",
    "    df = pd.read_csv('pathway_orphacode_hposet.csv')\n",
    "    \n",
    "    # Ensure n and m are within the specified limits\n",
    "    n = min(n, 1184)\n",
    "    m = min(m, 1410)\n",
    "    \n",
    "    # Extract first n rows and last m rows\n",
    "    rows_extracted = pd.concat([df.head(n), df.tail(m)])\n",
    "    \n",
    "    # Save the result to a txt file\n",
    "    rows_extracted.to_csv('pathway_phenotypes.csv')\n",
    "    \n",
    "    return rows_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c5244cb4-a5f4-4413-8b91-c00e15024391",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Function to randomly select HPO terms to simulate different scenarios for phenotype simulation\n",
    "def pathway_hpo(i, n, m):\n",
    "\n",
    "    \"\"\"\n",
    "    Simulate user-defined scenarios by selecting a specified number of HPO terms related to rare diseases \n",
    "    for phenotype simulation.\n",
    "\n",
    "    Args:\n",
    "        i (int): A non-negative integer specifying the number of rare disease-related HPO terms to select, \n",
    "        simulating scenarios where 'i' rare disease-related HPO terms are randomly chosen for phenotype simulation.  \n",
    "        \n",
    "        n (int): A non-negative integer specifying the number of rare disease-related HPO terms to select, \n",
    "        simulating scenarios where 'n' rare disease-related HPO terms are randomly chosen. Additionally, \n",
    "        'm' HPO terms are randomly selected from the entire set of HPO terms to introduce noise for phenotype simulation.\n",
    "        \n",
    "        m (int): A non-negative integer specifying the number of rare disease-unrelated HPO terms to select, \n",
    "        simulating scenarios where 'n' rare disease-related HPO terms are randomly chosen. Additionally, \n",
    "        'm' HPO terms are randomly selected from the entire set of HPO terms to introduce noise for phenotype simulation.\n",
    "\n",
    "    Returns:\n",
    "        None: This function generates and saves a csv file:\n",
    "        `pathway_hpo.csv`: Contains the simulated HPO terms user selected for each sample.\n",
    "        \n",
    "    \"\"\"    \n",
    "        \n",
    "    # read the orphanet code and hpoid of pathway information\n",
    "    df_pathway_hposet = pd.read_csv('pathway_phenotypes.csv')\n",
    "    \n",
    "    def random_hpo(hpo_str, i):\n",
    "        # Split the HPO terms by \"; \" and randomly select n terms\n",
    "        hpo_terms = hpo_str.split(\"; \")\n",
    "        selected_terms = random.sample(hpo_terms, min(i, len(hpo_terms)))  # limit to available terms\n",
    "        return \"; \".join(selected_terms)\n",
    "    \n",
    "    # Function to select random HPO terms from a string\n",
    "    def random_hpo_noise(hpo_str, num_terms):\n",
    "        # Split the HPO terms by \"; \" and randomly select the specified number of terms\n",
    "        hpo_terms = hpo_str.split(\"; \")\n",
    "        selected_terms = random.sample(hpo_terms, min(num_terms, len(hpo_terms)))  # limit to available terms\n",
    "        return selected_terms\n",
    "   \n",
    "    # read all the hpo term\n",
    "    df_all_hpo = pd.read_csv('hpo_id_all.csv')\n",
    "    \n",
    "    # Convert df_all_hpo 'hpo_id' column to a list of all HPO terms\n",
    "    all_hpo_terms = df_all_hpo['hpo_id'].tolist()\n",
    "    \n",
    "    # Apply the function to create the HPO_NOISE column in df_pathway_hposet\n",
    "    def combine_hpo_noise(row):\n",
    "        # Select n terms from the HPO_ALL column in df_pathway_hposet\n",
    "        selected_hpo_all = random_hpo_noise(row['HPO_ALL'], n)\n",
    "        # Select m terms from the df_all_hpo HPO list\n",
    "        selected_hpo_noise = random.sample(all_hpo_terms, min(m, len(all_hpo_terms)))\n",
    "        # Combine and return as a single string with \"; \" delimiter\n",
    "        return \"; \".join(selected_hpo_all + selected_hpo_noise)\n",
    "\n",
    "    # Apply the function to each row in df_pathway_hposet\n",
    "    df_pathway_hposet[f'HPO_{i}'] = df_pathway_hposet['HPO_ALL'].apply(lambda x: random_hpo(x, i))    \n",
    "\n",
    "    # Add the new column HPO_NOISE to df_pathway_hposet\n",
    "    df_pathway_hposet['HPO_NOISE'] = df_pathway_hposet.apply(combine_hpo_noise, axis=1)\n",
    "    \n",
    "    # Save the result to a csv file\n",
    "    df_pathway_hposet.to_csv('pathway_hpo.csv')\n",
    "    \n",
    "    return df_pathway_hposet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5488c67-60c3-4142-8b8f-cb2fc5b6f612",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
